sstables/sstable_set.cc

```diff
@@ -1118,18 +1118,25 @@ time_series_sstable_set::create_single_key_sstable_reader(
                 std::move(create_reader), std::move(filter), *pos.key(), schema, permit, fwd_sm, reversed));
 }
 
-compound_sstable_set::compound_sstable_set(schema_ptr schema, std::vector<sstable_set> sets)
+compound_sstable_set::compound_sstable_set(schema_ptr schema, std::vector<sstable_set> cloned_sets)
+    : _schema(std::move(schema))
+    , _cloned_sets(std::move(cloned_sets))
+{
+    _sets = boost::copy_range<std::vector<const sstable_set*>>(_cloned_sets | boost::adaptors::transformed([] (const sstable_set& sst_set) {
+        return &sst_set;
+    }));
+}
+
+compound_sstable_set::compound_sstable_set(schema_ptr schema, std::vector<const sstable_set*> sets)
     : _schema(std::move(schema))
     , _sets(std::move(sets)) {
 }
 
 sstable_set_impl_ptr compound_sstable_set::clone() const {
-    std::vector<sstable_set> cloned_sets;
-    cloned_sets.reserve(_sets.size());
-    for (const auto& set : _sets) {
-        cloned_sets.emplace_back(set.clone());
-    }
-    return seastar::make_shared<compound_sstable_set>(_schema, std::move(cloned_sets));
+    std::vector<sstable_set> cloned_sets = boost::copy_range<std::vector<sstable_set>>(_sets | boost::adaptors::transformed([] (const sstable_set* sst_set_ptr) {
+        return sst_set_ptr->clone();
+    }));
+    return seastar::make_shared<compound_sstable_set>(compound_sstable_set(_schema, std::move(cloned_sets)));
```

Comment:
Does clone of compound set need really to invoke underlying sstable set's clone?

If we say that sstable_set_impl is still immutable, then compound set's clone only need to create a new sstable set with a copy of _impl (not clone full content), as we do for range scans now.

Reply:
Maybe it's safer this way as compound set could be managing compound sets, and we want it work with any depth of compound sets.